hal/tests/TESTS/pin_names/generic/CMakeLists.txt

```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2021 ARM Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+include(mbed_greentea)
+
+# This test is enabled via CMake command-line argument -DMBED_HAL_PIN_NAMES_GENERIC_TEST=ON.
+if(NOT MBED_HAL_PIN_NAMES_GENERIC_TEST)
+    set(TEST_SKIPPED "Pin names generic test is not enabled for this Target")
+endif()
```

Comment:
> @ARMmbed/mbed-os-core As per the PR description, this test can't be built or run if a target has no LED or button. Such check is not available to CMake, because pins are defined in a header file. As we okay with the idea of an additional CMake variable?
> 
> If we want to do this,
> 
>    * Available options need to be documented in a place that's easier for users to see. I don't know if [CMake: Update README.md with CTest/greentea instructions #15001](https://github.com/ARMmbed/mbed-os/pull/15001) is a suitable place?
> 
>    * We can add `option(MBED_HAL_PIN_NAMES_GENERIC_TEST "Enable pin names generic test" OFF)` above, for readability.

I had a look at this with @rajkan01 yesterday. Initially I suggested setting [SKIP_REGULAR_EXPRESSION](https://cmake.org/cmake/help/latest/prop_test/SKIP_REGULAR_EXPRESSION.html) and emitting a trace message from the test to set the skipped state. However, @rajkan01 ran into a problem with this approach and it didn't work as expected. In that case an additional variable seemed reasonable (another option would be to add another field to targets.json to say "has_button" or similar, which didn't seem ideal at the time either). 

Anyway, now I see the code, I'm not entirely convinced that forcing the user to set this is the best approach. One potential issue: how will CI know when to pass this cache variable? Maybe we have a solution for that, I don't know. 

At this point I'm wondering if we should add an extra field to targets.json, so we can automate skipping/enabling this test. It's not the cleanest solution but it might be the best we can do. It wouldn't really be "test specific" config, as we're describing the available peripherals on the device. This particular test only runs if LED or BUTTON pins are available. https://github.com/ARMmbed/mbed-os/blob/b851ed92d29eaab7199211eed570db63217c0d0c/hal/tests/TESTS/pin_names/generic/main.cpp#L27

If we don't want to add an extra field to targets.json, and we decide the current approach is valid, i.e we want to add extra "user-facing" variables, we should of course document this clearly. We should also add the option so it shows up in ccmake and cmake-gui.